// @joshwooding/vite-plugin-react-docgen-typescript/dist/__tests__/generateDocgenCodeBlock.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const vitest_1 = require("vitest");
const parser_js_1 = require("react-docgen-typescript/lib/parser.js");
const generateDocgenCodeBlock_1 = require("../generateDocgenCodeBlock");
function getGeneratorOptions(parserOptions = {}) {
    return (filename) => {
        const filePath = path_1.default.resolve(__dirname, "__fixtures__", filename);
        return {
            filename,
            source: fs_1.default.readFileSync(filePath, "utf8"),
            componentDocs: (0, parser_js_1.parse)(filePath, parserOptions),
            docgenCollectionName: null,
            setDisplayName: true,
            typePropName: "type",
        };
    };
}
function loadFixtureTests() {
    return fs_1.default
        .readdirSync(path_1.default.resolve(__dirname, "__fixtures__"))
        .map(getGeneratorOptions());
}
const fixtureTests = loadFixtureTests();
const simpleFixture = fixtureTests.find((f) => f.filename === "Simple.tsx");
(0, vitest_1.describe)("component fixture", () => {
    fixtureTests.forEach((generatorOptions) => {
        (0, vitest_1.it)(`${generatorOptions.filename} has code block generated`, () => {
            (0, vitest_1.expect)((0, generateDocgenCodeBlock_1.generateDocgenCodeBlock)(generatorOptions)).toMatchSnapshot();
        });
    });
});
(0, vitest_1.it)("adds component to docgen collection", () => {
    (0, vitest_1.expect)((0, generateDocgenCodeBlock_1.generateDocgenCodeBlock)({
        ...simpleFixture,
        docgenCollectionName: "STORYBOOK_REACT_CLASSES",
    })).toMatchSnapshot();
});
(0, vitest_1.it)("generates value info for enums", () => {
    (0, vitest_1.expect)((0, generateDocgenCodeBlock_1.generateDocgenCodeBlock)(getGeneratorOptions({ shouldExtractLiteralValuesFromEnum: true })("DefaultPropValue.tsx"))).toMatchSnapshot();
});
